include/zephyr/bluetooth/hci_types.h

```diff
@@ -755,7 +755,7 @@ struct bt_hci_handle_count {
 #define BT_HCI_OP_HOST_NUM_COMPLETED_PACKETS    BT_OP(BT_OGF_BASEBAND, 0x0035) /* 0x0c35 */
 struct bt_hci_cp_host_num_completed_packets {
 	uint8_t  num_handles;
-	struct bt_hci_handle_count h[0];
+	FLEXIBLE_ARRAY_DECLARE(struct bt_hci_handle_count, h);
```

Comment:
Hi,

why not simply use foo[] instead of macro which adds some extra __unused_foo members under the hood?

also,  this declares unpacked structures inside packed structure, are those guaranteed to not affect packed struct alignment or no padding requirements?

Reply:
@sjanc this was also my worry (see my earlier comment). I'd prefer if the macro was only used in the corner cases where you can't directly use `[]`.

Reply:
I can do that, the macro though is safer since it ensures that these structures will work even if they are added to a union (which happens in some cases). I was in doubt about which approach to take to be honest.

Regarding padding / alignment it does not change anything because the size of this structure is 0. I have tested it an unless I have missed something there is no change regarding padding / alignment.

I think in the end the discussion is about which format to use. If you guys prefer just to use `[]` I will change the places where it is possible. What I need is to get rid of `[0]` .

Reply:
I prefer the standard `[]`. `FLEXIBLE_ARRAY_DECLARE` uses an empty struct, which is not standard C.

This file `hci_types.h` is ostensibly a C / C++ polyglot (it has `#ifdef __cplusplus`). In C++ an empty struct has size 1, giving `FLEXIBLE_ARRAY_DECLARE` a different meaning in C and C++.

The anonymous struct is not packed and the array is not the first member of the struct. This means the padding is implementation defined. This is probably not an issue in practice, but we should not rely on this without test coverage.

Reply:
That is a valid point, I will limit the usage of the macro to only it is really needed.